week_1/Array/move_zeroes/move_zeroes.js

```diff
@@ -0,0 +1,22 @@
+// Given an integer array nums, this algorithm moves all 0's to the end of it while maintaining the relative order of the non-zero elements.
+
+// This operation occurs in place, meaning that the array is mutated. Be careful!
+/**
+ * @param {number[]} nums
+ * @return {void} Do not return anything, modify nums in-place instead.
+ */
+const moveZeroes = (nums) => {
+	let zeroCount = 0;
+	for (i = 0; i < nums.length; i++) {
+		if (nums[i] === 0) {
```

Comment:
I think this solution would be much easier to read if it didn't attempt to mutate each index in place, but rather the entire array.

```js
let zeroCount = 0;
let filteredNumbers = nums.filter(num => num === 0);
// combine the filter numbers with a new array that full of zeros
nums = [...filteredNumbers, ...Array(zeroCount).fill(0)]
```

It would be interesting to see a comparison of performance between the two solutions here. I'd wager your original solution is more performant, but it's a bit difficult to comprehend.

Reply:
I see now that a requirement for this issue was to do this "in place".

I'm beginning to become a bit concerned that these requirements are encouraging bad practices. These type of requirements are less problematic for languages with stronger type systems, but very scary to see in JavaScript (zero type safety).